### PR TITLE
[Feature] Support TSM-MobileNetV2

### PR DESCRIPTION
**Pull request by irvingzhang0512**
## TODO list
+ [x] mobilenetv2 backbone & unittest.
+ [x] tsm-mobilenetv2 backbone & unittest.
+ [x] covnert checkpoint from origin repo.
  + original repo: 30 test crops, 19520 samples, top1/5 accuracy is 69.54%/88.66%
  + mmaction2 convertion: 10 test crops, 18219 samples, top1/5 accuracy is 69.04%/88.23%.
+ [x] Refactor mobilenet with mmcls
+ [x] changelog
+ [x] training with mmaction2 & update model zoo.
  + I don't have enough gpus to train on kinetics400, maybe next week i can have a try...
  + 贫穷的泪水

training results of mobilenet-tsm with DenseSampleFrames1x1x8. (origin ckpt get 69.54%/88.66% top1/5 accuracy).

|lr|epochs|gpus|weight decay|top1/5 acuracy|
|:-:|:-:|:-:|:-:|:-:|
|0.00875|50|7|0.0001|63.75%/85.52%|
|0.0025|50|4|0.0001|65.11%/85.99%|
|0.0025|100|4|0.0001|66.xx%/86.xx%|
|0.004|100|4|0.00004|68.31%/88.00%|
|0.0075|100|6|0.00004|68.41%/88.07%|

### CODE REVIEW

**Comment by dreamerlin** [comment]
What a contribution ! We can help to train some config settings to share your gpu pressure : )

**Comment by irvingzhang0512** [comment]
> What a contribution ! We can help to train some config settings to share your gpu pressure : )

The original repo doesn't provide training configs for tsm-mobilenetv2, so the training configs are copied from mmaction2-tsm-resnet50. Not sure if it is ok.

**Comment by irvingzhang0512** [comment]
trained mobilenet-tsm with `DenseSampleFrames`1x1x8/0.00875lr/7 gpus, get top1/5 acuracy 63.75%/85.52%, far bellow origin ckpt 69.54%/88.66%.
May have a try in 2 weeks, a little busy these days.

File: mmaction/models/backbones/mobilenetv2.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
+import torch.nn as nn
+from mmcv.cnn import constant_init, kaiming_init
+
+
+def conv_bn(inp, oup, stride):
```
Comment (kennymckormick) [inline mmaction/models/backbones/mobilenetv2.py:5]: Can we use ConvModule in mmcv instead?

File: mmaction/models/backbones/mobilenetv2.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        """
+        super(InvertedResidual, self).__init__()
+        self.stride = stride
+        assert stride in [1, 2]
+
+        hidden_dim = int(inp * expand_ratio)
+        self.use_res_connect = self.stride == 1 and inp == oup
+
+        if expand_ratio == 1:
+            self.conv = nn.Sequential(
```
Comment (kennymckormick) [inline mmaction/models/backbones/mobilenetv2.py:57]: Better to use 2 ConvModule rather than a long sequential

File: mmaction/models/backbones/mobilenetv2.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+
+class MobileNetV2(nn.Module):
+
+    def __init__(self,
+                 width_mult=1.,
+                 inverted_residual_setting=None,
+                 round_nearest=8,
+                 block=None,
+                 pretrained=False):
```
Comment (kennymckormick) [inline mmaction/models/backbones/mobilenetv2.py:109]: Better to make pretrained an url or None, instead of a bool. Since you may use different pretrain models.

**Comment by irvingzhang0512** [comment]
@kennymckormick next week I will refactor mobilenet with mmclassification

**Comment by kennymckormick** [comment]
> @kennymckormick next week I will refactor mobilenet with mmclassification

Great! It would be the best case if we can directly use mmcls's mobilenet.

**Comment by irvingzhang0512** [comment]
trained mobilenet-tsm with DenseSampleFrames1x1x8/0.0025lr/4 gpus/50epochs, get top1/5 acuracy 65.11%/85.99%, far bellow origin ckpt 69.54%/88.66%(100 epochs). The best ckpt is from the 50th epoch, I guess more epochs would help. 

@dreamerlin Maybe when I finish refactor mobilenetv2 with mmcls, you could kindly help to train this model with 8 gpus and 100 epochs, and upload the ckpt/json/log

**Comment by irvingzhang0512** [comment]
@dreamerlin ready to train and review.

Successfully convert origin ckpt to mmcls_mobilenetv2_tsm, get 69.5% top 1 val accuracy.

**Comment by irvingzhang0512** [comment]
trained mobilenet-tsm with DenseSampleFrames1x1x8/0.004lr/4 gpus/100epochs/0.00004 weight decay, get top1/5 acuracy 68.31%/88.00%, much better than before, not far away from origin ckpt 69.54%/88.66%(100 epochs). 

Maybe the remaining 1% gap is because my kinetics-400 dataset only have 220k videos for training set & 18k for val set.

**Comment by dreamerlin** [comment]
> trained mobilenet-tsm with DenseSampleFrames1x1x8/0.004lr/4 gpus/100epochs/0.00004 weight decay, get top1/5 acuracy 68.31%/88.00%, much better than before, not far away from origin ckpt 69.54%/88.66%(100 epochs).
> 
> Maybe the remaining 1% gap is caused by the lack of training samples(There are only 220k videos in my kinetics-400 dataset training set)

Sorry for my late respond due to personal affair. trained mobilenet-tsm with DenseSampleFrames1x1x8/0.01lr/8 gpus/100epochs/0.00004 weight decay, get top1/5 acuracy 68.33%/88.34%. Any suggestion ? @innerlee @irvingzhang0512

**Comment by irvingzhang0512** [comment]
I'll double check network architecture first in two days.

**Comment by irvingzhang0512** [comment]
@dreamerlin Do you use `ThreeCrop` during testing? In original tsm repo, mobilenetv2-tsm was tested with 30 crops for each sample. The configs I commit only use CenterCrop in test pipeline.

**Comment by dreamerlin** [comment]
> @dreamerlin Do you use `ThreeCrop` during testing? In original tsm repo, mobilenetv2-tsm was tested with 30 crops for each sample. The configs I commit only use CenterCrop in test pipeline.

Test with `ThreeCrop` and full size (256, 256)

**Comment by irvingzhang0512** [comment]
It seems that accuracy for training set is not high enough, so maybe we chould use smaller weight decay, say 2e-5.

I try to train/test on new version of kinetics-400, downloaded from [here](https://academictorrents.com/details/184d11318372f70018cf9a72ef867e2fb9ce1d26)(240618 training samples & 19404 val samples). 

Train with DenseSampleFrames1x1x8/0.0075lr/6 gpus/100epochs/0.00002, Test with `dict(type='ThreeCrop', crop_size=256)`. 

|model|top1/5 accuracy|
|:-:|:-:|
|paper|69.54%/88.66%|
|convert from origin tsm ckpt|71.15%/89.54%|
|train by myself|69.98%/89.04%|

I run though my codes and dataset again and again, but no bugs found... But I don't believe these, the results are too good...

Could you help to verify these on your dataset? @dreamerlin 

FYI, ckpt convertion script

```python
import torch
from mmcv import Config
from mmaction.models import build_model

# load origin tsm models
# https://file.lzhu.me/projects/tsm/models/TSM_kinetics_RGB_mobilenetv2_shift8_blockres_avg_segment8_e100_dense.pth
d = torch.load(
    "/path/to/TSM_kinetics_RGB_mobilenetv2_shift8_blockres_avg_segment8_e100_dense.pth"
)['state_dict']
origin_list = list(d.keys())
origin_list = [
    e for e in origin_list
    if 'num_batches_tracked' not in e and 'running' not in e
]

# create mmaction tsm-mobilenetv2 model
cfg_path = "/path/to/tsm_mobilenetv2_video_inference_dense_1x1x8_100e_kinetics400_rgb.py"
cfg = Config.fromfile(cfg_path)
model = build_model(cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg)
target_list = list(model.named_parameters())

# convert ckpt
transform_dict = {}
for e1, e2 in zip(origin_list, target_list):
    assert d[e1].shape == e2[1].shape
    assert e1.split(".")[-1] == e2[0].split(".")[-1]
    transform_dict[e1[:e1.rfind(".")]] = e2[0][:e2[0].rfind(".")]
model.load_state_dict(
    {
        transform_dict[ori[:ori.rfind(".")]] + "." + ori.split(".")[-1]: d[ori]
        for ori in d.keys()
    },
    strict=False)

# save ckpt
torch.save({
    'state_dict': model.state_dict()
}, './path/to/tsm_mobilenetv2_converted.pth')
```

**Comment by dreamerlin** [comment]
> It seems that accuracy for training set is not high enough, so maybe we could use smaller weight decay, say 2e-5.
> 
> I try to train/test on new version of kinetics-400, downloaded from [here](https://academictorrents.com/details/184d11318372f70018cf9a72ef867e2fb9ce1d26)(240618 training samples & 19404 val samples).
> 
> Train with DenseSampleFrames1x1x8/0.0075lr/6 gpus/100epochs/0.00002, Test with `dict(type='ThreeCrop', crop_size=256)`.
> 
> model	top1/5 accuracy
> paper	69.54%/88.66%
> convert from origin tsm ckpt	71.15%/89.54%
> train by myself	69.98%/89.04%
> I went though my codes and dataset again and again, no bugs found... But I don't believe these, the results are too good...
> 
> Could you help to verify these on your dataset? @dreamerlin
> 
> FYI, ckpt convertion script
> 
> ```python
> import torch
> from mmcv import Config
> from mmaction.models import build_model
> 
> # load origin tsm models
> # https://file.lzhu.me/projects/tsm/models/TSM_kinetics_RGB_mobilenetv2_shift8_blockres_avg_segment8_e100_dense.pth
> d = torch.load(
>     "/path/to/TSM_kinetics_RGB_mobilenetv2_shift8_blockres_avg_segment8_e100_dense.pth"
> )['state_dict']
> origin_list = list(d.keys())
> origin_list = [
>     e for e in origin_list
>     if 'num_batches_tracked' not in e and 'running' not in e
> ]
> 
> # create mmaction tsm-mobilenetv2 model
> cfg_path = "/path/to/tsm_mobilenetv2_video_inference_dense_1x1x8_100e_kinetics400_rgb.py"
> cfg = Config.fromfile(cfg_path)
> model = build_model(cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg)
> target_list = list(model.named_parameters())
> 
> # convert ckpt
> transform_dict = {}
> for e1, e2 in zip(origin_list, target_list):
>     assert d[e1].shape == e2[1].shape
>     assert e1.split(".")[-1] == e2[0].split(".")[-1]
>     transform_dict[e1[:e1.rfind(".")]] = e2[0][:e2[0].rfind(".")]
> model.load_state_dict(
>     {
>         transform_dict[ori[:ori.rfind(".")]] + "." + ori.split(".")[-1]: d[ori]
>         for ori in d.keys()
>     },
>     strict=False)
> 
> # save ckpt
> torch.save({
>     'state_dict': model.state_dict()
> }, './path/to/tsm_mobilenetv2_converted.pth')
> ```

Thanks~ I will try it

**Comment by dreamerlin** [comment]
got top1 69.69 from the converted ckpt on my dataset. Maybe you can provide your trained ckpt,

**Comment by irvingzhang0512** [comment]
https://wp.cethik.com:443/link/9BE08E853DE84DAF716C7E1156C5C9A4

**Comment by dreamerlin** [comment]
> https://wp.cethik.com:443/link/9BE08E853DE84DAF716C7E1156C5C9A4

got 68.39 for this :P

**Comment by irvingzhang0512** [comment]
> > https://wp.cethik.com:443/link/9BE08E853DE84DAF716C7E1156C5C9A4
> 
> got 68.39 for this :P Both use short-side 320p frames

In my dataset, get val 69.04% with center crop 224 only...Different trainint/val set, quite different results. Could you please share your filelist(train&val). I'd like to check the difference between our datasets...

**Comment by dreamerlin** [comment]
> > > https://wp.cethik.com:443/link/9BE08E853DE84DAF716C7E1156C5C9A4
> > 
> > 
> > got 68.39 for this :P Both use short-side 320p frames
> 
> In my dataset, get val 69.04% with center crop 224 only...Different trainint/val set, quite different results. Could you please share your filelist(train&val). I'd like to check the difference between our datasets...

[kinetics_train_list_hd320.txt](https://github.com/open-mmlab/mmaction2/files/5878033/kinetics_train_list_hd320.txt)
[kinetics_val_list_hd320.txt](https://github.com/open-mmlab/mmaction2/files/5878037/kinetics_val_list_hd320.txt)

**Comment by irvingzhang0512** [comment]
The number of our datasets is almost the same(training set 240436/240618, val set 19796/19404).
However, 6-7%(1000 samples) of val set and 3-4%(9000 samples) of training set are different.
@dreamerlin Could you please train this model with weight decay 2e-5? It seems that pre-trained models trained by dataset is not suitable for your dataset...

**Comment by irvingzhang0512** [comment]
@innerlee @dreamerlin  Now the models I trained get better accuracy than original tsm repo(69.54% -> 69.98%). This is probably becuase datasets are quite different. I don't know what I should do now... Any suggestions?

**Comment by kennymckormick** [comment]
@dreamerlin Have you trained TSM w. MobileNetv2 with the original repo on our datasets, maybe this number can be used for reference.

**Comment by innerlee** [comment]
If accuracies are good, then it's good to go.

Please resolve the conflicts.

@dreamerlin Could you reproduce the result? Can you upload the checkpoints?

File: docs/changelog.md [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 - Support Spatio-Temporal Action Detection(AVA) ([#351](https://github.com/open-mmlab/mmaction2/pull/351))
 - Support to return feature maps in `inference_recognizer` ([#458](https://github.com/open-mmlab/mmaction2/pull/458))
+- Support TSM-MobileNetV2. ([#415](https://github.com/open-mmlab/mmaction2/pull/415))
```
Comment (innerlee) [inline docs/changelog.md:11]: This line should go to a newer verison

File: mmaction/models/backbones/mobilenet_v2.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+                    norm_cfg=self.norm_cfg,
+                    act_cfg=self.act_cfg,
+                    with_cp=self.with_cp))
+            self.in_channels = out_channels
+
+        return nn.Sequential(*layers)
+
+    def init_weights(self):
+        if isinstance(self.pretrained, str):
+            logger = logging.getLogger()
```
Comment (innerlee) [inline mmaction/models/backbones/mobilenet_v2.py:256]: @dreamerlin double check whether the get_root_logger should be used here

File: mmaction/models/backbones/mobilenet_v2.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                 pretrained=None,
+                 widen_factor=1.,
+                 out_indices=(7, ),
+                 frozen_stages=-1,
+                 conv_cfg=dict(type='Conv'),
+                 norm_cfg=dict(type='BN2d', requires_grad=True),
+                 act_cfg=dict(type='ReLU6', inplace=True),
+                 norm_eval=False,
+                 with_cp=False):
+        super(MobileNetV2, self).__init__()
```
Comment (innerlee) [inline mmaction/models/backbones/mobilenet_v2.py:163]: `super().__init__()`

File: mmaction/models/backbones/mobilenet_v2.py (hunk 5 above)
Comment (irvingzhang0512) [reply mmaction/models/backbones/mobilenet_v2.py:256]: mobilenet_v2 is directly copied from mmcls. I'll modify these codes to 

```python
from ...utils import get_root_logger()

logger = get_root_logger()
```

**Comment by dreamerlin** [comment]
> @dreamerlin Have you trained TSM w. MobileNetv2 with the original repo on our datasets, maybe this number can be used for reference.

Yeah, I can help to train it :P

**Review by innerlee — APPROVED** [review]

**Comment by dreamerlin** [comment]
> > @dreamerlin Have you trained TSM w. MobileNetv2 with the original repo on our datasets, maybe this number can be used for reference.
> 
> Yeah, I can help to train it :P

top1: 68.46, using 3 crop and full size (256)

**Comment by irvingzhang0512** [comment]
Don't know what to do. Maybe we should follow @kennymckormick 's suggestion, train tsm-mobilenetv2 with the original repo.

I'm a little busy these days, may have a try after spring festival..

**Comment by kennymckormick** [comment]
@innerlee I think we can merge this PR once the performance is the same or better than the paper. performance alignment with the released ckpt can be postponed, what's ur opinion?

**Comment by innerlee** [comment]
Yeah better than paper is good enough

**Comment by kennymckormick** [comment]
@dreamerlin Would you please help add the ckpts you trained? Maybe it's OK to merge it first and try to improve the performance later.

**Comment by dreamerlin** [comment]
related config files and links are updated now

**Comment by innerlee** [comment]
failing tests are related

**Review by kennymckormick — APPROVED** [review]

**Comment by innerlee** [comment]
Thanks!
